test/helper.py

```diff
@@ -66,8 +66,14 @@ def run(command, *, timeout=30):
     logger = logger_from_command(command)
     logger.info('running: %s', command)
 
-    proc = subprocess.run(shlex.split(command), capture_output=True, text=True, check=False,
-                          timeout=timeout)
+    try:
+        proc = subprocess.run(shlex.split(command), capture_output=True, text=True, check=False,
+                              timeout=timeout)
+    except subprocess.TimeoutExpired as e:
+        for line in e.output.splitlines():
+            if line:
+                logger.info('stdout: %s', line)
+        raise
```

Comment:
Shouldn't `stderr` then be printed as well?

Might it be an option to assign stoud and stderr to variables and handle printout in a `finally` clause?

Reply:
Yeah, printing `stderr` makes sense. I'm not sure I'm getting the point with `finally` block. You mean to handle printout in `finally` _instead_ of `except` block?

Reply:
I ment combining the output for the good case and the error case, yes.

Something like

``` diff
diff --git a/test/helper.py b/test/helper.py
index b067d6e..232aa5c 100644
--- a/test/helper.py
+++ b/test/helper.py
@@ -66,15 +66,21 @@ def run(command, *, timeout=30):
     logger = logger_from_command(command)
     logger.info('running: %s', command)
 
-    proc = subprocess.run(shlex.split(command), capture_output=True, text=True, check=False,
-                          timeout=timeout)
-
-    for line in proc.stdout.splitlines():
-        if line:
-            logger.info('stdout: %s', line)
-    for line in proc.stderr.splitlines():
-        if line:
-            logger.warning('stderr: %s', line)
+    try:
+        proc = subprocess.run(shlex.split(command), capture_output=True, text=True, check=False,
+                              timeout=timeout)
+        stdout = proc.stdout
+        stderr = proc.stderr
+    except subprocess.TimeoutExpired as e:
+        stdout = e.stdout
+        stderr = e.stderr
+    finally:
+        for line in stdout.splitlines():
+            if line:
+                logger.info('stdout: %s', line)
+        for line in stderr.splitlines():
+            if line:
+                logger.warning('stderr: %s', line)
 
     logger.info('exitcode: %d', proc.returncode)
```
But I have no clue how this actually differs from just printing the error itself and this solution also wouldn't allow to easily re-raise the error.

@Bastian-Krause do you have any better idea maybe?

Reply:
Published slightly modified version of your proposal